### PR TITLE
trading_index() align

### PR DESCRIPTION
**Pull request by ValueRaider**
So consequence of shifting XTAE start time to 9:59am (#247) is it also shifts trading_index() back a minute. Annoying because some data providers don't do this (e.g. Yahoo), instead they shift back further to nicely align data depending on interval. 

```
import yfinance as yf
yf.Ticker("ESLT.TA").history(period="1d", interval="15m").index[0]
```
> Timestamp('2022-10-25 09:45:00+0300', tz='Asia/Jerusalem')

So this PR replicates that behaviour via an optional `align` argument. Implementation just shifts open times ~back~ forwards/backwards in `_TradingIndex.__init__()`, direction depending on sign (+ve/-ve)

Before:
```
cal.trading_index(start="2022-09-04",end="2022-09-05", period="30m", intervals=True, force_close=True)
```
> IntervalIndex([[2022-09-04 06:59:00, 2022-09-04 07:29:00), ... [2022-09-05 13:59:00, 2022-09-05 14:15:00)], dtype='interval[datetime64[ns, UTC], left]')

After, with alignment:
```
cal.trading_index(start="2022-09-04",end="2022-09-05", period="30m", intervals=True, force_close=True, align='-30m')
```
> IntervalIndex([[2022-09-04 06:30:00, 2022-09-04 07:00:00), ... [2022-09-05 14:00:00, 2022-09-05 14:15:00)], dtype='interval[datetime64[ns, UTC], left]')

### CODE REVIEW

**Comment by ValueRaider** [comment]
When #254 is merged, I'll "rebase" this then tests should pass.

**Comment by ValueRaider** [comment]
When #254 is merged I'll rebase then test will pass (requires the corrected XTAE open time).

**Comment by maread99** [comment]
@ValueRaider, #254 is now merged although I don't think you need to bother rebasing this one. There are no apparent conflicts and tests are passing. Leave it with me, I hope to look at it this week.

**Comment by ValueRaider** [comment]
Ah, if tests are run after merging (in a venv) then they should pass.

**Review by maread99 — COMMENTED** [review]
Hi @ValueRaider.

Thanks for this PR. This functionality is absolutely required.

I think you're approach to implement it by way of adjusting the opens is the way to do it, and I like the option to re-anchor the left of each session's first indice either forwards or backwards.

I've proposed some changes via a review.

Also, I feel there should be provision for aligning the afternoon subsessions of those calendars with breaks. Although there's currently no pre-defined calendar that could require this functionality, it seems reasonable to include it (test will probably require a fictional calendar). For the time being, could just assume that any alignment should be applied to both the am and pm subsessions (rather than providing for these to be aligned independently).

I know you alluded to it, but I hadn't appreciated you were working on a branch that doesn't include the commits that were made changes to XTAE. **Would you mind squashing the commits here and opening a new PR based on the current master** (I'd like to play around with it locally and the check tests are still passing)?

Cheers!

File: exchange_calendars/exchange_calendar.py [hunk 1]
```diff
@@ -2314,6 +2314,7 @@ def trading_index(
         curtail_overlaps: bool = False,
         ignore_breaks: bool = False,
         parse: bool = True,
+        align: pd.Timedelta | str | None = None,
```
[the remark this answers is not in the document]
Comment (ValueRaider) [reply exchange_calendars/exchange_calendar.py:2317]: Agree that end feels wrong, I didn't want to break existing codes.

**Comment by ValueRaider** [comment]
> Also, I feel there should be provision for aligning the afternoon subsessions of those calendars with breaks.

Only thing I'll push back against. I think better to wait for a real-world scenario requiring this before implementing.
- maybe scenario never happens because they're always aligned to hour
- if happens, then maybe user would want the afternoon alignment to be independent of morning, e.g. morning align back and afternoon align forwards.

> Would you mind squashing the commits here and opening a new PR based on the current master

Sure, once discussions here resolved.

**Comment by maread99** [comment]
>Only thing I'll push back against. I think better to wait for a real-world scenario requiring this before implementing

Hi, yeah, no worries, I would like it covered, for custom calendars more than anything, although I appreciate there's no pre-defined calendar at the moment that requires it - if you could squash and rebase and then I'll add a commit to cover the pm subsessions.

Cheers

**Comment by ValueRaider** [comment]
I've squashed and rebased, but kept this PR as it seems to be OK with changes.

**Comment by maread99** [comment]
Thanks @ValueRaider.

I'm struggling for time this week. If I don't get a chance over the weekend then I hope to add a commit no later than next week.

**Comment by maread99** [comment]
Hi @ValueRaider

I notice the test only covers the the case where `align` takes the same absolute value as `period`. I agree this is the most common use case, although if they can diverge then this circumstance needs to be tested for. So we either:
* Change `align` to take `Literal['previous', 'next', None]` with the effect of aligning backwards / forwards respectively to the `period` (or by default, `None`, doing nothing).
* Expand the test to cover cases where `align` is not the same as `period`.

Personally I'd favor the former. It would somewhat simplify the implementation and I'd question if there's a use case for differing periods / alignments(?) Let me know what you think.

Separately, in any event the testing needs to be extended to:
* cover all of "XLON", "XHKG", "CMES", "24/7" to make sure these still return as expected (i.e. unchanged) when an alignment is passed.
* verify the full index returns as expected, rather than just the left side of the first interval.

One approach might be to:
* use the `calendars` fixture to parametrize the test.
* for each calendar, create indexes with and without alignment and then 'test the difference' between them is as expected.

**Comment by ValueRaider** [comment]
I prefer the former but it must be the latter because Yahoo aligns hourly data back to 9:30. Leave this with me

**Comment by maread99** [comment]
Annoying. But at least serves to make clear that there's no knowing how a provider will decide to consolidate the data, we just need to offer the flexibility to accommodate it.

The commit I just added implements 'align-pm'.

If you could look at extending the tests as mentioned earlier, then I'll sort out the parsing test which is failing (as the TODO by the test) and add a test to check 'align_pm' against a fictional calendar.

Cheers

**Comment by ValueRaider** [comment]
I've expanded the test, next will extend to other calendars.

I need to clarify this comment:

> I notice the test only covers the the case where `align` takes the same absolute value as `period`

This is wrong. I create a new `periods` array using keys from `aligned_start_times` then iterate through both for all permutations.

**Comment by maread99** [comment]
I've included `align` and `align_pm` to the fuzz tests.

I'd like to leave `test_align` in some form to serve as a hard-coded concrete test. I still need to include a fictional calendar with an unaligned break-end to more fully test `align_pm`. Also still needs to test effect of `align` when passing `start` and `end` with time components as opposed to as sessions.

**Comment by ValueRaider** [comment]
I've partially parameterized `test_align`. Still hardcoded is aligned start times, I'm not sure that can be parameterized.

> a session of "us_futures" covers two dates

Can you provide an example? I had a quick look at "XCBF" and didn't see any examples.

**Comment by maread99** [comment]
> Can you provide an example?

Sessions of the "CMES" calendar or the fictional calendar named "us_futures".

> Still hardcoded is aligned_start_times

That's grand, we'll use `test_align` as a concrete test against known expected aligned start times. The changes to the fuzz tests should cover the rest.

I've amended `test_align`:
* to look at specific sessions of a modified test calendar
* so that hypothesis also defines the parameters `align`, `align_pm`, `ignore_breaks` and the force options. 

I realised it would be possible for `align` or `align_pm` to cause indices to overlap that wouldn't otherwise (amended the overlaps errors' messages and added a test).

It's my intention tomorrow to squash, merge and cut a release. Let me know if you want me to hold off for any reason.

**Comment by ValueRaider** [comment]
> It's my intention tomorrow to squash, merge and cut a release. Let me know if you want me to hold off for any reason.

I'll just reply to this part for now because it's late for me - go ahead.

**Comment by ValueRaider** [comment]
> It's my intention tomorrow to squash, merge and cut a release. Let me know if you want me to hold off for any reason.

Go ahead

**Comment by maread99** [comment]
Done.

Thanks raising this one @ValueRaider.

**Pull request merged**